lua/compe/config.lua

```diff
@@ -20,27 +22,51 @@ Config.setup = function(config)
   config.allow_prefix_unmatch = Config._true(config.allow_prefix_unmatch)
 
   -- normalize source metadata
-  for name, metadata in pairs(config.source) do
-    if type(metadata) ~= 'table' then
-      config.source[name] = { disabled = not Config._true(metadata) }
-    else
-      config.source[name].disabled = config.source[name].disabled or false
+  if config.source then
+    for name, metadata in pairs(config.source) do
+      if type(metadata) ~= 'table' then
+        config.source[name] = { disabled = not Config._true(metadata) }
+      else
+        config.source[name].disabled = config.source[name].disabled or false
+      end
     end
+  else
+    config.source = {}
   end
 
-  Config._config = config
+  return config
+end
+
+Config.setup = function(config, bufnr)
+  config = Config._normalize(config)
+
+  if bufnr == nil then
+    Config._config = config
+  else
+    for key, value in pairs(Config._config) do
+      if config[key] == nil then
+        config[key] = value
+      end
+    end
```

Comment:
It is a shallow merge.

So suggested use-case will be realized by the below codes.

```viml
lua <<EOF
local compe = require"compe"

-- global configuration.
compe.setup {
  source = {
    necosyntax = true,
    buffer = true,
    path = true,
  }
}

local on_attach = function()
  -- Disable necosyntax source when nvim_lsp is activated.
  compe.setup({
    source = {
      nvim_lsp = true,
      buffer = true,
      path = true,
    }
  }, 0)
end
require"lspconfig".vimls.setup {
  on_attach = on_attach
}
EOF
```

Is it enough for @ii14?

Reply:
Almost, the merging doesn't work. I'm trying to figure out what's wrong.